extension/src/openvic-extension/singletons/CursorSingleton.cpp

```diff
@@ -0,0 +1,585 @@
+	OV_BIND_METHOD(CursorSingleton::get_sequence,{"cursor name"},"normal");
+	OV_BIND_METHOD(CursorSingleton::get_resolutions,{"cursor name"},"normal");
+	OV_BIND_METHOD(CursorSingleton::generate_resolution,{"cursor name","base resolution index", "target resolution"},"normal",0,Vector2i(64,64));
+	OV_BIND_METHOD(CursorSingleton::get_cursor_names);
+
+	ADD_PROPERTY(PropertyInfo(
+		Variant::ARRAY,
+		"cursor_names", PROPERTY_HINT_ARRAY_TYPE,
+		"String"),
+	"", "get_cursor_names");
+}
+
+CursorSingleton* CursorSingleton::get_singleton() {
+	return _singleton;
+}
+
+CursorSingleton::CursorSingleton() {
+	ERR_FAIL_COND(_singleton != nullptr);
+	_singleton = this;
+}
+
+CursorSingleton::~CursorSingleton() {
+	ERR_FAIL_COND(_singleton != this);
+	_singleton = nullptr;
+}
+
+
+Array CursorSingleton::get_frames(String const& name, int res_index){
+	if(res_index < cursors[name].images.size()){
+		return cursors[name].images[res_index];
+	}
+	return Array();
+}
+
+TypedArray<Vector2i> CursorSingleton::get_hotspots(String const& name, int res_index){
+	if(res_index < cursors[name].hotspots.size()){
+		return cursors[name].hotspots[res_index];
+	}
+	return TypedArray<Vector2i>();
+}
+
+int CursorSingleton::get_animationLength(String const& name){
+	return cursors[name].animationLength;
+}
+
+TypedArray<Vector2i> CursorSingleton::get_resolutions(String const& name){
+	return cursors[name].resolutions;
+}
+
+PackedFloat32Array CursorSingleton::get_displayRates(String const& name){
+	if(cursors[name].displayRates.has_value()){
+		return cursors[name].displayRates.value();
+	}
+	return PackedFloat32Array();
+}
+
+PackedInt32Array CursorSingleton::get_sequence(String const& name){
+	if(cursors[name].sequence.has_value()){
+		return cursors[name].sequence.value();
+	}
+	return PackedInt32Array();
+}
+
+void CursorSingleton::generate_resolution(String const& name, int base_res_index, Vector2i target_res){
+	cursor_asset_t* cursor = &cursors[name];
+	Array new_frameset;
+	TypedArray<Vector2i> new_hotspots;
+
+	for(int i=0;i<cursor->images[base_res_index].size();i++){
+
+		Ref<ImageTexture> texture = static_cast<Ref<ImageTexture>>(cursor->images[base_res_index][i]);
+		
+		Ref<Image> image;
+		image.instantiate();
+		image->create(target_res.x, target_res.y, false, Image::Format::FORMAT_RGBA8);
+		image->copy_from(texture->get_image());
+		image->resize(target_res.x,target_res.y,Image::INTERPOLATE_BILINEAR);
+
+
+		Ref<ImageTexture> new_texture = Ref<ImageTexture>();
+		new_texture.instantiate();
+		new_texture->set_image(image);
+
+		new_frameset.push_back(new_texture);
+
+		Vector2i base_hotspot = cursor->hotspots[base_res_index][i];
+		Vector2i scale_ratio = target_res / static_cast<Vector2i>(cursor->resolutions[base_res_index]);
+		new_hotspots.push_back(base_hotspot*scale_ratio);
+
+	}
+
+	cursor->images.push_back(new_frameset);
+	cursor->hotspots.push_back(new_hotspots);
+	cursor->resolutions.push_back(target_res);
+}
```

Comment:
```suggestion
void CursorSingleton::generate_resolution(StringName const& name, int32_t base_res_index, Vector2i target_res) {
	cursor_map_t::iterator it = cursors.find(name);
	ERR_FAIL_COND_MSG(it == cursors.end(), vformat("Cursor \"%s\" not found", name));

	cursor_asset_t& cursor = it.value();

	ERR_FAIL_INDEX_MSG(
		base_res_index, cursor.images.size(), vformat("Invalid image index for cursor \"%s\": %d", name, base_res_index)
	);

	TypedArray<ImageTexture> const& images = cursor.images[base_res_index];
	TypedArray<Vector2i> const& hotspots = cursor.hotspots[base_res_index];
	const Vector2i resolution = cursor.resolutions[base_res_index];

	TypedArray<ImageTexture> new_frameset;
	TypedArray<Vector2i> new_hotspots;

	for (size_t index = 0; index < images.size(); index++) {
		Ref<ImageTexture> const& texture = images[index];

		Ref<Image> image;
		image.instantiate();
		image->copy_from(texture->get_image());
		image->resize(target_res.x, target_res.y, Image::INTERPOLATE_BILINEAR);

		new_frameset.push_back(ImageTexture::create_from_image(image));

		new_hotspots.push_back(static_cast<Vector2i>(hotspots[index]) * target_res / resolution);
	}

	cursor.images.push_back(new_frameset);
	cursor.hotspots.push_back(new_hotspots);
	cursor.resolutions.push_back(target_res);
}
```
- `StringName`, `int32_t` and `TypedArray<ImageTexture>` as mentioned in header comments.
- `find` for looking up in `cursors` to avoid crash on missing entry. The local `cursor` variable can be a reference rather than a pointer as it'll never be null. We need to use `it.value()` here rather than `it->second` as we need a non-const reference (it's a weird quirk of the map containers we use, when accessed as a `(key, value)` pair they're always const to protect the key from being modified, but we can still get the non-const value via the `it.value()` helper function).
- Error check on `base_res_index` (we only need to check against one of the vectors, assuming images, hotspots and resolutions are always grown in tandem).
- Local `images`, `hotspots` and `resolution` variables to avoid unnecessary extra lookups in each iteration of the for loop.
- Better to use `size_t index` as it matches the type of `images.size()` and it's generally better to avoid single-letter variable names, both for clarity and to avoid typos.
- There's no need to call `image->create`, `copy_from` overrides any existing data within the image (we still need `instantiate` to allocate memory for the `Image`'s internal variables before they can then be set by `copy_from`).
- We can use the static helper function `ImageTexture::create_from_image` rather than manually instantiating the `ImageTexture` and setting its `Image`.
- Better to do `static_cast<Vector2i>(hotspots[index]) * target_res / resolution` rather than `static_cast<Vector2i>(hotspots[index]) * (target_res / resolution), it reduces the integer division error, e.g. `6 * 4 / 3 = 24 / 3 = 8` while `6 * (4 / 3) = 6 * 1 = 6`. The `static_cast<Vector2i>` wouldn't be necessary if we were using a `Packed...Array`, maybe we could use `PackedVector2Array` but that might require some extra code to ensure the new hotspot is made up of whole numbers, e.g. by using `Vector2`'s `floor` function.

Reply:
Since hotspots was changed to a PackedVector2Array, I got rid of the cast, and added .floor() as suggested.